rfcs/20220307-monitoring_across_the_stack.md

```diff
@@ -0,0 +1,353 @@
+4. Run the local monitoring stack to ensure that the saved configurations in a brand new deployment match the expected configurations from staging.
+    * This may require exercising the local service a bit to check that metrics appear as expected in the dashboards... can be difficult with small amounts of data though.
+5. Deploy them to production via terraform or some other method to update the running Prometheus and Grafana instances with the latest configurations
+    * Terraform has [providers for uploading dashboard configurations to  Grafana](https://registry.terraform.io/providers/grafana/grafana/latest)
+    * With Prometheus, generate the new configuration file per environment using `just mkpromconf [environment]` and upload the configuration to the environment then restart Prometheus.
+
+## DevEx
+
+I've added some new `just` recipes in the example code in this PR.
+
+Grafana configurations are stored as JSON so I think they'd be linted by prettier. They should be copy/paste exports anyway so we don't need to lint them for correctness or even style technically.
+
+Prometheus configurations can be linted by `promtool` using this pre-commit repo: https://github.com/fortman/pre-commit-prometheus.
+
+## Outstanding questions
+
+### Authentication
+
+Do we want to hide the metrics endpoint behind authentication? I don't really see why we would other than to stop spamming (but we could throttle those endpoints to match the prometheus scrape rate for example and it would prevent anyone from doing too much damage).
+
+I like the idea of having fully open Prometheus and Grafana dashboards that don't require authentication to _explore_. Obviously we'd want to hide administration inside of authentication but each Prometheus and Grafana handle that on their own.
+
+### Alerting
+
+Do we want to use Prometheus for alerting or Grafana?
+
+Pros for Prometheus:
+- It is possible and potentially even trivial to provision alerts based on configuration files
+Cons for Prometheus:
+- It is much harder to configure and automatically provision changes to Prometheus's configuration
+- The UI for building them does not exist so I'm not sure what the best way to write new rules is other than via yaml files which will be harder for folks to contribute to given we're not all full-time devops people with tons of time to learn how to do this without fiddling around with visual helper tools
+
+Pros for Grafana:
+- The UI for building them is nice
+Cons for Grafana:
+- There's no way to export or provision them so the only way to add rules is through the UI (there's no JSON version of the form either). Moving rules across environments would be hard... to be fair, developing rules with anything except production data is also very difficult (lots of guessing) so maybe this isn't a problem.
+
+I'd rather use Grafana provided we back up the database of alert configurations. If we do that then not having them in the code is hopefully less of a disaster.
```

Comment:
Same, I'd much rather try to fiddle with this in the UI than do an iterative process _only for production_ where we have to edit config files.

Reply:
Same for me as well.